### PR TITLE
Async bindings for producers and consumers

### PR DESCRIPTION
**Pull request by Leonidas-from-XIV**
After merging in all parts of the preparatory work, here's the main bit: the Async bindings.

### CODE REVIEW

**Comment by didier-wenzek** [comment]
Thank you for this new pull request to support Async. I will try to help, even if not used to Async!

Did you forget to commit some files? The async_tests are referencing the function `new_producer_topic` which is not defined.

**Comment by Leonidas-from-XIV** [comment]
Oops, indeed. I've renamed `new_producer_topic` to `new_topic` in the end, since the consumer does not need to use `Kafka.topic` handles at all, the API uses the topic as a string in any case.

I'm currently trying to get the integration test to run only in the `Kafka_async` package and was told by Rudi that there was some issue with this that was fixed in dune 2.3 which is why I am trying to build with that and see if it helps.

**Comment by didier-wenzek** [comment]
Just to confirm that I got the same error as Travis on my Unbuntu 18.04 with OCaml 4.09

```
$ make integration
dune build --only-packages kafka,kafka_async,kafka_lwt @integration --force
 integration alias test_async/integration/integration (got signal SEGV)
```

**Comment by Leonidas-from-XIV** [comment]
Ah yes, true, I was developing #17 and this in parallel and forgot to rebase this code on top of that once it was merged.

File: test_async/consumer/dune [hunk 1]
```diff
@@ -0,0 +1,4 @@
+(executable
+  (name consumer)
```
Comment (didier-wenzek) [inline test_async/consumer/dune]: I would also add

```
  (public_name kafka-async-consumer)
  (package kafka_async)
```

because it's convenient to have command-line tools even if redundant with those provided by Kafka.

File: test_async/producer/dune [hunk 2]
```diff
@@ -0,0 +1,4 @@
+(executable
+  (name producer)
```
Comment (didier-wenzek) [inline test_async/producer/dune]: I would also add

```
  (public_name kafka-async-producer)
  (package kafka_async)
```

**Comment by didier-wenzek** [comment]
It's great to have `Async` consumer and producer!

There is a strange issue though.
* The integration tests (`_build/default/test_async/integration/integration.exe`) produce the expected message but then hang.
* The test producer works fine, but the test consumer misses the messages if the integration tests are blocked. The messages are properly produced (tested with another consumer) but the async consumer sees nothing.
* The messages are produced and consumed properly by the async tools if the integration tests are not running.

**Comment by didier-wenzek** [comment]
Sorry, the previous errors make sense. I failed to notice that the integration tests and the consumer share the same `group.id`.

File: lib_async/kafka_async.mli [hunk 3]
```diff
@@ -0,0 +1,28 @@
+open Async
+
+type producer
+
+type consumer
+
+type 'a response = ('a, Kafka.error * string) result
+
+val produce :
+  producer ->
```
Comment (didier-wenzek) [inline lib_async/kafka_async.mli]: I would expect something more consistent with the Lwt and Sync interfaces. These two don't need a producer. Idem for `rd_kafka_produce`. Why is this required for Async?

File: lib_async/kafka_async.mli [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+  unit response Deferred.t
+
+val new_producer : (string * string) list -> producer response
+
+val new_consumer : (string * string) list -> consumer response
+
+val new_topic :
+  producer -> string -> (string * string) list -> Kafka.topic response
+
+val consume : consumer -> topic:string -> Kafka.message Pipe.Reader.t response
```
Comment (didier-wenzek) [inline lib_async/kafka_async.mli]: Same remark as for the `produce` function. Besides, the topic should be a value not a string. Otherwise, how to pass options to the topic.

File: lib_async/kafka_async.mli [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+  ?key:string ->
+  string ->
+  unit response Deferred.t
+
+val new_producer : (string * string) list -> producer response
+
+val new_consumer : (string * string) list -> consumer response
+
+val new_topic :
+  producer -> string -> (string * string) list -> Kafka.topic response
```
Comment (didier-wenzek) [inline lib_async/kafka_async.mli]: Maybe a type `handler`, wrapping either a `producer` or a `consumer`, might be useful here.

We can also discuss the sync and Lwt interfaces. It can be helpful to distinguish producers and consumers rather than having only handlers.

File: lib_async/kafka_async.mli (hunk 3 above)
Comment (Leonidas-from-XIV) [reply lib_async/kafka_async.mli]: I don't particularly like the idea of having mutable state on the module level (where `pending_msg` is in the Lwt bindings) so I opted to move it to the `producer` type, so multiple producers are independent and GCing a producer also cleans out the table.

File: lib_async/kafka_async.mli (hunk 4 above)
Comment (Leonidas-from-XIV) [reply lib_async/kafka_async.mli]: I used to use a `Kafka.topic` here, but the type is not all that useful, since it contains a `handle` (`rd_kafka_t` which is already in the `consumer`) and a `rd_kafka_topic_t` but the `rd_kafka_subscribe` API never uses `rd_kafka_topic_t`, instead it uses a `rd_kafka_topic_partition_list_t` which is constructed from topic names as a string. No options can be passed to topics, so the only thing it did was to get the name of the topic as string. So there was no need to build a `Kafka.topic` and indeed creating one made it more confusing because one would expect the options to do something, but rdkafka would never see the options.

In the end I removed the `Kafka.topic` type from here since the type system can't certify that the `handle` in the `consumer` matches the `handle` in the `Kafka.topic` and the code would then just ignore one of these which is somewhat unintiutive (e.g. a `Kafka.topic` for a completely different connection would work just as well, since the code would only ever retrieve the name of the connection as string).

File: lib_async/kafka_async.mli (hunk 5 above)
Comment (Leonidas-from-XIV) [reply lib_async/kafka_async.mli]: I agree. In the beginning I had just one handler type, but looking at the way handlers for consumers & producers are created in rdkafka they are distinct (either `RD_KAFKA_PRODUCER` or `RD_KAFKA_CONSUMER`) and I don't think e.g. using a producer for consuming would work. The C code just kinda ignores it and just calls both of them `rd_kafka_t`, thus allowing to misuse the API.

So I thought it would be nice to avoid this potential pitfall (given that we probably want a program that compiles to be somewhat correct) to use a phantom type to distinguish them on the OCaml side. But the more I worked on it, the more the contents of the OCaml types diverged to a point where I thought that making them distinct, unrelated types is better. This helped me to avoid making the `producer` and `consumer` types polymorphic but it unfortunately has lead to a bit of duplication of code. That can probably be cleaned up by creating a common subset of these types, a e.g. a `connection` type which would contain the `start`, `stop` and `handle` keys.

It also unfortunately means that producers and consumers have to be closed using different functions, but I judged the added type safety is worth it.

**Comment by Leonidas-from-XIV** [comment]
Right, that's a good point! I changed the default `group.id` in the consumer and made it overrideable. I should probably also add options to set the broker list, given these binaries will be shipped with the library, then hardcoding `localhost:9092` is a bit limiting.

**Comment by Leonidas-from-XIV** [comment]
The tests currently all fail because of a ocaml-ci-scripts failure, I've submitted a [pull request](https://github.com/ocaml/ocaml-ci-scripts/pull/330) to hopefully solve this.

File: lib_async/kafka_async.mli (hunk 3 above)
Comment (didier-wenzek) [reply lib_async/kafka_async.mli]: Good points. I will have a look to see what can be done with the Lwt and Sync interfaces.

File: lib_async/kafka_async.mli (hunk 5 above)
Comment (didier-wenzek) [reply lib_async/kafka_async.mli]: I agree: the benefits of typed handlers worth the value, even if there is some duplicate code.

File: lib_async/kafka_async.mli (hunk 4 above)
Comment (didier-wenzek) [reply lib_async/kafka_async.mli]: I don't follow here. Topics have configuration [options](https://github.com/edenhill/librdkafka/blob/master/CONFIGURATION.md#topic-configuration-properties). Besides, even if just a string is used behind the scene, I think that it's valuable to have a typed topic value.

The second point about two `handle`s is more concerning. Yes, this would be a bad design, possibly leading to code working for strange reasons. Maybe, the OCaml topic might be a wrapper around a (handler, topic) pair?

Comment (Leonidas-from-XIV) [reply lib_async/kafka_async.mli]: The thing is, the `rd_kafka_subscribe` C API does not take any options since nowhere in that chain it uses a `rd_kafka_topic_t`, so there is no way to pass these options to the C code, so librdkafka would not know about these options.

In the existing code options on a topic are set using `rd_kafka_topic_conf_t` and then a `rd_kafka_topic_t` is created (which saves the options somewhere internally). Consuming is done (among others) via `rd_kafka_consume_start` which takes this `rd_kafka_topic_t` and can apply the options.

Contrast that with `rd_kafka_subscribe`, which takes a handle and a `rd_kafka_topic_partition_list_t`. That list is created from an empty list and topics are added via `rd_kafka_topic_partition_list_add` which is populated with `const char * topic`. It is unaware of any `rd_kafka_topic_t` and therefore does not have a reference to any `rd_kafka_topic_conf_t`. So requiring a `Kafka.handle` is strange since the options defined there would not be passed to rdkafka when subscribing.

I don't know either why exactly `rd_kafka_topic_partition_list_add` does not just take a `rd_kafka_topic_t` or how options for that balanced group are supposed to be set (the only option that can be set is `int32_t partition`). I can reach out to the rdkafka people and ask why the balanced group subscriptions do not accept any options.

**Comment by Leonidas-from-XIV** [comment]
Hmm, it looks like there is a [default setting](https://github.com/edenhill/librdkafka/wiki/FAQ#how-do-i-set-topic-configuration-properties-with-the-kafkaconsumer) for consumer topics and it seems like the way to set custom topics is to indeed create topics in the consumer and then just don't use the `rd_kafka_topic`.

**Comment by Leonidas-from-XIV** [comment]
I've pushed a commit which uses `Kafka.topic`, but I am somewhat unhappy with the API, since the `Kafka.topic` and the `consumer` have different handles and might even be on completely different connections which is not great. A nicer way would be to use the connection handle from the topic, but we need the metadata that is stored in the `consumer`.

My best idea here would be a `Kafka_async.consumer_topic` type, which contains a reference to the consumer and the type, thus preventing users from passing two unrelated arguments to `subscribe` but I am open to other suggestions.

**Comment by didier-wenzek** [comment]
Having a `Kafka_async.consumer_topic` seems to be a good idea for the same reason it's good to have distinct types for `consumer` and `producer`.

This would remove the need for 2 parameters and the risks of passing a topic from one context to another.

I will give a try to that idea on the sync and Lwt interface, with four specific types for consumer/producer variants of handler/topic. This will be an opportunity to lift the list of pending ids out of the binding into the module itself, as you suggested for good reasons.

**Comment by Leonidas-from-XIV** [comment]
(Note that currently the builds fail because the CI scripts were changed to [use 4.09.1 instead of 4.09.0](https://github.com/ocaml/ocaml-ci-scripts/pull/329) but the cached archive uses 4.09.0 which confuses `opam` and causes it to abort. Removing the caches in Travis solves this particular issue)
